File: app/src/main/java/com/sumian/sd/diary/sleeprecord/calendar/custom/SleepCalendarViewVH.java
package com.sumian.sd.diary.sleeprecord.calendar.custom;

import android.content.Context;
import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.sumian.sd.R;
import com.sumian.sd.diary.sleeprecord.calendar.calendarView.CalendarViewVH;

/**
 * <pre>
 *     @author : Zhan Xuzhao
 *     e-mail : xuzhao.z@sumian.com
 *     time   : 2018/5/29 15:59
 *     desc   :
 *     version: 1.0
 * </pre>
 */
public class SleepCalendarViewVH extends CalendarViewVH {
    private SleepCalendarViewVH(View itemView) {
        super(itemView);
    }

    @NonNull
    public static CalendarViewVH create(ViewGroup parent) {
        Context context = parent.getContext();
        View inflate = LayoutInflater.from(context).inflate(R.layout.item_calendar, parent, false);
        return new SleepCalendarViewVH(inflate);
    }

    @Override
    public void setDay(int day, int dayType) {
        String text = day > 0 ? String.valueOf(day) : "";
        getMTextView().setText(text);
        getMTextView().setTextColor(getTextColor(dayType));
        getMTextView().setBackground(getBgDrawable(dayType));
        getMTextView().setTypeface(Typeface.DEFAULT, isBold(dayType) ? Typeface.BOLD : Typeface.NORMAL);
    }

    private boolean isBold(int dayType) {
        return false;
    }

    @Override
    protected int getTextColor(int dayType) {
        int textColor;
        switch (dayType) {
            case SleepDayType.HAS_DATA:
            case SleepDayType.NO_DATA:
                textColor = R.color.t1_color;
                break;
            case SleepDayType.SELECT_HAS_DATA:
            case SleepDayType.SELECT_NO_DATA:
                textColor = R.color.white;
                break;
            case SleepDayType.FEATURE:
                textColor = R.color.t1_color_40;
                break;
            default:
                textColor = R.color.t1_color;
                break;
        }
        return getMContext().getResources().getColor(textColor);
    }

    @Override
    protected Drawable getBgDrawable(int dayType) {
        int drawableRes;
        switch (dayType) {
            case SleepDayType.HAS_DATA:
                drawableRes = R.drawable.ic_calendar_date;
                break;
            case SleepDayType.SELECT_HAS_DATA:
                drawableRes = R.drawable.ic_calendar_selecteddate;
                break;
            case SleepDayType.SELECT_NO_DATA:
                drawableRes = R.drawable.ic_calendar_selected;
                break;
            default:
                return null;
        }
        return getMContext().getResources().getDrawable(drawableRes);
    }
}

File: app/src/main/java/com/sumian/sd/diary/sleeprecord/calendar/custom/SleepCalendarViewVH.kt
package com.sumian.sd.diary.sleeprecord.calendar.custom

import android.graphics.Typeface
import android.graphics.drawable.Drawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.sumian.common.utils.ColorCompatUtil
import com.sumian.sd.R
import com.sumian.sd.diary.sleeprecord.calendar.calendarView.CalendarViewVH

@Suppress("DEPRECATION")
/**
 * <pre>
 * @author : Zhan Xuzhao
 * e-mail : xuzhao.z@sumian.com
 * time   : 2018/5/29 15:59
 * desc   :
 * version: 1.0
</pre> *
 */
class SleepCalendarViewVH private constructor(itemView: View) : CalendarViewVH(itemView) {

    override fun setDay(day: Int, dayType: Int, secondType: Int) {
        val text = if (day > 0) day.toString() else ""
        mTextView.text = text
        mTextView.setTextColor(getTextColor(dayType, secondType))
        mTextView.background = getBgDrawable(dayType)
        mTextView.setTypeface(Typeface.DEFAULT, if (isBold(dayType)) Typeface.BOLD else Typeface.NORMAL)
        mViewBg.background = getSecondBgDrawable(secondType)
    }

    @Suppress("UNUSED_PARAMETER")
    private fun isBold(dayType: Int): Boolean {
        return false
    }

    override fun getTextColor(dayType: Int, secondType: Int): Int {
        val textColor: Int = if (secondType == SECOND_BG_TYPE_NONE) {
            when (dayType) {
                SleepDayType.HAS_DATA, SleepDayType.NO_DATA -> R.color.t1_color
                SleepDayType.SELECT_HAS_DATA, SleepDayType.SELECT_NO_DATA -> R.color.white
                SleepDayType.FEATURE -> R.color.t1_color_40
                else -> R.color.t1_color
            }
        } else {
            R.color.white
        }
        return ColorCompatUtil.getColor(mContext, textColor)
    }

    override fun getBgDrawable(dayType: Int): Drawable? {
        val drawableRes: Int
        when (dayType) {
            SleepDayType.HAS_DATA -> drawableRes = R.drawable.ic_calendar_date
            SleepDayType.SELECT_HAS_DATA -> drawableRes = R.drawable.ic_calendar_selecteddate
            SleepDayType.SELECT_NO_DATA -> drawableRes = R.drawable.ic_calendar_selected
            else -> return null
        }
        return mContext.resources.getDrawable(drawableRes)
    }

    private fun getSecondBgDrawable(secondBgType: Int): Drawable? {
        val resId = when (secondBgType) {
            SECOND_BG_TYPE_START -> R.drawable.shape_calendar_week_start_bg
            SECOND_BG_TYPE_MIDDLE -> R.drawable.shape_calendar_week_middle_bg
            SECOND_BG_TYPE_END -> R.drawable.shape_calendar_week_end_bg
            SECOND_BG_TYPE_START_END -> R.drawable.shape_calendar_week_start_end_bg
            else -> 0
        }
        return if (resId == 0) {
            null
        } else {
            mContext.resources.getDrawable(resId)
        }
    }

    companion object {
        @JvmStatic
        val SECOND_BG_TYPE_NONE = 0
        @JvmStatic
        val SECOND_BG_TYPE_START = 1
        @JvmStatic
        val SECOND_BG_TYPE_MIDDLE = 2
        @JvmStatic
        val SECOND_BG_TYPE_END = 3
        @JvmStatic
        val SECOND_BG_TYPE_START_END = 4


        fun create(parent: ViewGroup): CalendarViewVH {
            val context = parent.context
            val inflate = LayoutInflater.from(context).inflate(R.layout.item_calendar, parent, false)
            return SleepCalendarViewVH(inflate)
        }
    }
}
